pg-extend/src/pg_alloc.rs

```diff
@@ -7,23 +7,89 @@
 
 //! A Postgres Allocator
 
-use std::alloc::{GlobalAlloc, Layout};
+use std::alloc::{self, GlobalAlloc, Layout};
+use std::cell::Cell;
 use std::ffi::c_void;
+use std::sync::{Mutex, Once};
 
 use crate::pg_sys;
 
+// static INIT_LOCK: Once = Once::new();
+// static mut LOCK: Option<Mutex<()>> = None;
+
+thread_local! {
+    static IN_PGCONTEXT: Cell<bool> = Cell::new(false);
+}
+
 /// An allocattor which uses the palloc and pfree functions available from Postgres.
-/// 
+///
 /// This is managed by Postgres and guarantees that all memory is freed after a transaction completes.
 pub struct PgAllocator;
 
+impl PgAllocator {
+    //     /// Initializes the PgAllocator, this must be called at least once before Rust code is executed;
+    //     pub fn init() {
+    //         INIT_LOCK.call_once(|| unsafe {
+    //             LOCK = Some(Mutex::new(()));
+    //         })
+    //     }
+}
+
+/// Initializes the PgAllocator, this must be called at least once before Rust code is executed;
+    pub fn set_in_pgcontext(b: bool) {
+        // INIT_LOCK.call_once(|| unsafe {
+        //     LOCK = Some(Mutex::new(()));
+        // })
+
+        IN_PGCONTEXT.with(|in_pgcontext| in_pgcontext.set(b));
+    }
+
+    pub fn is_in_pgcontext() -> bool {
+        IN_PGCONTEXT.with(Cell::get)
+    }
+
 unsafe impl GlobalAlloc for PgAllocator {
     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-        // TODO: is there anything we need ot do in terms of layout, etc?
-        pg_sys::palloc(layout.size()) as *mut u8
+        // PgAllocator::init();
+
+        if IN_PGCONTEXT.with(Cell::get) {
```

Comment:
Uhh, this looks incredibly unsafe. If `palloc` gets used to allocate in the main thread, then that variable gets transferred to another Rust thread which tries to free it, that would crash.

I think a better idea would be to disable `PgAllocator` in threaded Rust extensions entirely; when returning results from a Rust function, that data can be copied into a `palloc`-allocated buffer.

Reply:
Agreed. This entire idea is unsafe.